components/brave_wallet/common/eth_sign_typed_data_helper_unittest.cc

```diff
@@ -177,6 +178,51 @@ TEST(EthSignedTypedDataHelperUnitTest, EncodedData) {
   EXPECT_EQ(encoded_person_v4, encoded_person_v3);
 }
 
+TEST(EthSignedTypedDataHelperUnitTest, InvalidEncodedData) {
+    const std::string data_json(R"({
+    "from":{"name":"Cow","wallet":"0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826"},
+    "to":{"name":"Bob","wallet":"0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB"},
+    "contents":"Hello, Bob!"
+    })");
+    
+    auto data_value =
+      base::JSONReader::Read(data_json, base::JSON_PARSE_CHROMIUM_EXTENSIONS |
+                                            base::JSON_ALLOW_TRAILING_COMMAS);
+    ASSERT_TRUE(data_value);
+    auto& data_dict = data_value->GetDict();
+
+    for (const std::string& invalid_json : {
+           R"({
+    "Domain": [
+        { "name": ["AStringArray", "String2"], "type": "string" }
+    ]})",
+           R"({
+    "Domain": [
+        { "name": 1234, "type": "uint2556" }
+    ]})",
+           R"({
+    "Domain": [
+        { "name": { "name": "name" }, "type": "string" }
+    ]})",
+           R"({
+    "Domain": [
+        { "name": "name", "type": 1234 }
+    ]})"}) {
+    SCOPED_TRACE(invalid_json);
+    auto invalid_value = base::JSONReader::Read(
+        invalid_json, base::JSON_PARSE_CHROMIUM_EXTENSIONS |
+                          base::JSON_ALLOW_TRAILING_COMMAS);
+    ASSERT_TRUE(invalid_value);
+    
+    std::unique_ptr<EthSignTypedDataHelper> invalid_types_helper =
+        EthSignTypedDataHelper::Create(invalid_value->GetDict().Clone(),
+                                       EthSignTypedDataHelper::Version::kV4);
+  auto encoded_domain_v4 =
+      invalid_types_helper->EncodeData("Domain", *(data_dict.FindDict("name")));
```

Comment:
This doesn't make sense, `data_dict.FindDict("name")` will always return `nullptr` and crash on dereferencing because you don't have name field at the top level of dictionary. And your data isn't matching type, for this test you should just use data like `R"({"name":"Cow"})"` and pass the dictionary directly.

Reply:
The reason you won't see it crash when running test is because the dereferencing happens after the check of 
`if(!type_str || !name_str) {return absl::nullopt;}` instead of on the spot of passing due to compiler optimization. And still we shouldn't pass the problematic data, the only invalid data we should pass is what we test other parts of the input data should be completely valid to isolate the scope we want to test.

Reply:
cleaned up in [`deb3723` (#17292)](https://github.com/brave/brave-core/pull/17292/commits/deb372334f8662fb82b271e1be9fa279e4976e13)